services/kvstore/src/main/java/org/hyperledger/besu/services/kvstore/InMemoryKeyValueStorage.java

```diff
@@ -77,9 +77,17 @@ public boolean containsKey(final byte[] key) throws StorageException {
 
   @Override
   public long removeAllKeysUnless(final Predicate<byte[]> retainCondition) throws StorageException {
-    long initialSize = hashValueStore.keySet().size();
-    hashValueStore.keySet().removeIf(key -> !retainCondition.test(key.getArrayUnsafe()));
-    return initialSize - hashValueStore.keySet().size();
+    long finalSize;
+    final Lock lock = rwLock.readLock();
+    lock.lock();
+    try {
+      long initialSize = hashValueStore.keySet().size();
+      hashValueStore.keySet().removeIf(key -> !retainCondition.test(key.getArrayUnsafe()));
+      finalSize = initialSize - hashValueStore.keySet().size();
+    } finally {
+      lock.unlock();
+    }
+    return finalSize;
```

Comment:
For readability I would move the declaration of finalSize and the return inside the try block.  It has the same net effect - happy path returns size and unlocks and any exception will also unlock.